neqo-http3/src/stream_type_reader.rs

```diff
@@ -6,57 +6,120 @@
 
 #![allow(clippy::module_name_repetitions)]
 
-use crate::{AppError, Http3StreamType, HttpRecvStream, ReceiveOutput, RecvStream, Res, ResetType};
-use neqo_common::{Decoder, IncrementalDecoderUint};
+use crate::{
+    AppError, Error, Http3StreamType, HttpRecvStream, NewStreamType, ReceiveOutput, RecvStream,
+    Res, ResetType,
+};
+use neqo_common::{qtrace, Decoder, IncrementalDecoderUint};
 use neqo_transport::Connection;
 
 #[derive(Debug)]
 pub enum NewStreamTypeReader {
-    Read {
+    ReadType {
+        push_stream_allowed: bool,
+        reader: IncrementalDecoderUint,
+        stream_id: u64,
+    },
+    ReadSessionId {
         reader: IncrementalDecoderUint,
         stream_id: u64,
     },
     Done,
 }
 
 impl NewStreamTypeReader {
-    pub fn new(stream_id: u64) -> Self {
-        NewStreamTypeReader::Read {
+    pub fn new(stream_id: u64, push_stream_allowed: bool) -> Self {
+        NewStreamTypeReader::ReadType {
+            push_stream_allowed,
             reader: IncrementalDecoderUint::default(),
             stream_id,
         }
     }
 
-    pub fn get_type(&mut self, conn: &mut Connection) -> Option<Http3StreamType> {
-        // On any error we will only close this stream!
-        loop {
-            if let NewStreamTypeReader::Read {
-                ref mut reader,
-                stream_id,
-            } = self
-            {
+    fn read(&mut self, conn: &mut Connection) -> Res<(Option<u64>, bool)> {
+        match self {
+            NewStreamTypeReader::ReadType {
+                reader, stream_id, ..
+            }
+            | NewStreamTypeReader::ReadSessionId { reader, stream_id } => loop {
```

Comment:
I think that you want to use `if let N.... = self {` here.  I completely missed the loop bit on my first read.